### PR TITLE
fix webGL label rendering bug

### PR DESCRIPTION
**Pull request by dabingnn**
Re: cocos-creator/fireball#3977

Changes proposed in this pull request:
- 

> Makes sure these boxes are checked before submitting your PR - thank you!
> - [X] If your pull request has gone "stale", you should **rebase** your work on top of the latest version of the upstream branch.
> - [X] If your commit history is full of small, unimportant commits (such as "fix pep8" or "update tests"), **squash** your commits down to a few, or one, discreet changesets before submitting a pull request.
> - To official teams:
>   - [X] Check that your javascript is following our [style guide](https://github.com/cocos-creator/fireball/blob/dev/.github/CONTRIBUTING.md) and end files with a newline
>   - [X] Document new code with comments in source code based on [API Docs](https://github.com/cocos-creator/fireball#api-docs)

@cocos-creator/engine-admins

### CODE REVIEW

File: cocos2d/core/label/CCSGLabelCanvasRenderCmd.js [hunk 1]
```diff
@@ -469,6 +469,20 @@
...
             this._labelContext.fillText(this._splitedStrings[i], startPosition.x, startPosition.y + i * lineHeight);
         }
+        //premultiply alpha for webGL
+        if(!(this instanceof _ccsg.Label.CanvasRenderCmd)) {
+            var labelCanvas = this._labelCanvas;
+            var imageData = this._labelContext.getImageData(0,0,labelCanvas.width, labelCanvas.height);
+            var data = imageData.data;
+            var displayOpacity = this._displayedOpacity / 255;
+            for(var i = 0; i < data.length; i += 4) {
+                var alpha = data[i+3] = (data[i+3] * displayOpacity ) | 0;
```
Comment (jareguo) [inline cocos2d/core/label/CCSGLabelCanvasRenderCmd.js]: 这里 alpha 直接 `/ 255` 就好了吧

Comment (jareguo) [inline cocos2d/core/label/CCSGLabelCanvasRenderCmd.js]: 好吧，没注意到 `data[i+3]` 也赋值了

**Comment by jareguo** [comment]
LGTM

**Comment by pandamicro** [comment]
由于这个方案损耗过高，需要换方案，可以重新检查 Cocos2d-html5 和 Creator 实现的差异，因为 html5 版本没有这个问题

File: cocos2d/core/label/CCSGLabelCanvasRenderCmd.js [hunk 2]
```diff
@@ -471,7 +471,7 @@
         }
 
         this._texture._textureLoaded = false;
-        this._texture.handleLoadedTexture();
+        this._texture.handleLoadedTexture(true);
```
Comment (pandamicro) [inline cocos2d/core/label/CCSGLabelCanvasRenderCmd.js]: 这个对于 Canvas 渲染模式也有影响吧，Canvas 渲染指令中应该要设置一次 globalAlpha

File: cocos2d/core/label/CCSGLabelWebGLRenderCmd.js [hunk 3]
```diff
@@ -105,7 +105,10 @@ proto.uploadData = function (f32buffer, ui32buffer, vertexDataOffset) {
 
     // Fill in vertex data with quad information (4 vertices for sprite)
     // Use 255 because color has been set when baking label
-    this._color[0] = ((this._displayedOpacity<<24) | (255<<16) | (255<<8) | 255);
+    // premultiplied alpha is used for labelTTF and system font
+    var opacity = this._displayedOpacity;
+    this._color[0] = ((opacity<<24) | (opacity<<16) | (opacity<<8) | opacity);
```
Comment (pandamicro) [inline cocos2d/core/label/CCSGLabelWebGLRenderCmd.js:110]: 预乘之后应该是改 blend function 为 ONE 和 ONE_MINUS_SRC_ALPHA 就可以了吧

**Comment by pandamicro** [comment]
LGTM

File: cocos2d/core/label/CCSGLabelCanvasRenderCmd.js [hunk 4]
```diff
@@ -469,6 +469,20 @@
             }
             this._labelContext.fillText(this._splitedStrings[i], startPosition.x, startPosition.y + i * lineHeight);
         }
+        //premultiply alpha for webGL
```
Comment (zilongshanren) [inline cocos2d/core/label/CCSGLabelCanvasRenderCmd.js]: 这个每帧都会被调用吗？ 这样很耗性能吧？如果图片的 image data 数据很多的话？

**Pull request merged**